syntax = "proto3";

package todo;

service TodoService {
    rpc CreateUser(User) returns (CreateUserResponse) {}
    rpc GetUser(User) returns (User) {}
    rpc CreateTodo(Todo) returns (CreateTodoResponse) {}
    rpc GetTodo(Todo) returns (Todo) {}
    rpc GetUserTodos(User) returns (TodoList) {}
    rpc DeleteTodo(Todo) returns (DeleteResponse) {}
    rpc DeleteUser(User) returns (DeleteResponse) {}
}

message User {
    int64 id = 1;
    string name = 2;
    string email = 3;
}

message UsersList {
    repeated User users = 1;
}

message GetUsers {}

message Todo {
    int64 id = 1;
    int64 user_id = 2;
    string details = 3;
    Status status = 4;

    enum Status {
        UNKNOWN_TODO_STATUS = 0;
        TODO = 1;
        WORKINPROGRESS = 2;
        COMPLETE= 3;
    }
}

message TodoList {
    repeated Todo todos = 1;
}

message ActionStatus {
    enum Status {
        UNKNOWN = 0;
        SUCCESS = 1;
        FAILED = 2;
    }
    
    Status status = 1;
    repeated string errors = 2;
}

message CreateUserResponse {
    ActionStatus status = 1;
    User user = 2;
}

message CreateTodoResponse {
    ActionStatus status = 1;
    Todo todo = 2;
}

message DeleteResponse {
    ActionStatus status = 1;
}

